File: RansomNote.py

"""
Basic DSA - Ransom Note
Return True if ransom_note can be constructed from magazine (each letter can be used once).
"""


def can_construct(ransom_note, magazine):
    """
    Count letters in magazine then consume for ransom_note.
    Time Complexity: O(n + m)
    Space Complexity: O(1) for fixed alphabet, otherwise O(k)
    """
    counts = {}
    for ch in magazine:
        counts[ch] = counts.get(ch, 0) + 1

    for ch in ransom_note:
        if counts.get(ch, 0) == 0:
            return False
        counts[ch] -= 1
    return True


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Ransom Note")
    print("=" * 60)

    print(can_construct("a", "b"))      # False
    print(can_construct("aa", "ab"))    # False
    print(can_construct("aa", "aab"))   # True

File: RansomNote.java
/**
 * Basic DSA Problems - One Fundamental Hashing Question
 * Simple and essential DSA problem for beginners
 */
public class RansomNote {

    // Problem 1: Ransom Note
    // Given two strings ransomNote and magazine, return true if ransomNote
    // can be constructed by using the letters from magazine.
    //
    // Each letter in magazine can only be used once.
    //
    // Example:
    // ransomNote = "aa", magazine = "aab" -> true

    /**
     * Frequency counting for lowercase letters
     * Time Complexity: O(n + m)
     * Space Complexity: O(1)
     */
    public static Boolean canConstruct(String ransomNote, String magazine) {
        if (ransomNote == null || magazine == null) {
            return null;
        }

        int[] freq = new int[26];
        for (int i = 0; i < magazine.length(); i++) {
            char c = magazine.charAt(i);
            if (c < 'a' || c > 'z') {
                return null;
            }
            freq[c - 'a']++;
        }

        for (int i = 0; i < ransomNote.length(); i++) {
            char c = ransomNote.charAt(i);
            if (c < 'a' || c > 'z') {
                return null;
            }
            if (freq[c - 'a'] == 0) {
                return false;
            }
            freq[c - 'a']--;
        }

        return true;
    }

    // Test cases
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem 1: Ransom Note");
        System.out.println("=".repeat(60));

        System.out.println("ransomNote = \"a\", magazine = \"b\" -> " + canConstruct("a", "b"));
        System.out.println("ransomNote = \"aa\", magazine = \"ab\" -> " + canConstruct("aa", "ab"));
        System.out.println("ransomNote = \"aa\", magazine = \"aab\" -> " + canConstruct("aa", "aab"));
        System.out.println("ransomNote = \"\", magazine = \"abc\" -> " + canConstruct("", "abc"));
        System.out.println("ransomNote = null, magazine = \"abc\" -> " + canConstruct(null, "abc"));
    }
}